pinot-broker/src/main/java/org/apache/pinot/broker/routing/RoutingManager.java

```diff
@@ -458,6 +461,35 @@ public TimeBoundaryInfo getTimeBoundaryInfo(String offlineTableName) {
     return timeBoundaryManager != null ? timeBoundaryManager.getTimeBoundaryInfo() : null;
   }
 
+  /**
+   * Fetches segments which replicas are all in ERROR state, put the result to noReplicaSegments set.
+   */
+  public void fetchSegmentsInErrorState(ExternalView externalView, Set<String> noReplicaSegments) {
+    Set<String> partitionSet = externalView.getPartitionSet();
+    for (String partitionName : partitionSet) {
+      Map<String, String> stateMap = externalView.getStateMap(partitionName);
+      int numReplicas = stateMap.size();
+      int errorCount = 0;
+      for (String segmentState : stateMap.values()) {
+        if (SegmentOnlineOfflineStateModel.ERROR.equals(segmentState)) {
+          errorCount++;
+        }
+      }
+      if (errorCount == numReplicas) {
+        noReplicaSegments.add(partitionName);
+      }
+    }
+  }
+
+  /**
+   * Checks whether the broker request matches with any segments which replicas are all in ERROR state.
+   */
+  public boolean containsNoReplicaSegments(BrokerRequest brokerRequest) {
```

Comment:
```suggestion
  public int containsNoReplicaSegments(BrokerRequest brokerRequest) {
```
Return the number of segments unavailable here